src/execution_algorithms/ltfb/regularized_evolution.cpp

```diff
@@ -0,0 +1,343 @@
+  const int trainer_id = comm.get_trainer_rank();
+  auto const step = ctxt.get_step();
+
+  std::vector<EvalType> sample_trainers(num_trainers);
+  if (comm.am_world_master()) {
+    for (unsigned int i = 0; i < num_trainers; i++)
+      sample_trainers[i] = i;
+    std::shuffle(sample_trainers.begin(),
+                 sample_trainers.end(),
+                 get_ltfb_generator());
+  }
+  comm.world_broadcast(comm.get_world_master(),
+                       sample_trainers.data(),
+                       num_trainers);
+
+  // If rank within first S, send score , else score = 0
+  auto it =
+    std::find(sample_trainers.begin(), sample_trainers.end(), trainer_id);
+
+  El::Int score = evaluate_model(m, ctxt, dc);
+  // If in sample, send true score
+  if (std::distance(sample_trainers.begin(), it) < m_sample_size) {
+    // score = evaluate_model(m, ctxt, dc);
+    if (comm.am_trainer_master())
+      std::cout << "Trainer " << trainer_id << " in sample" << std::endl;
+  }
+  // Else force score to 0
+  else {
+    // score = evaluate_model(m, ctxt, dc);
+    score = 0;
+  }
+
+  std::vector<EvalType> score_list(num_trainers);
+  comm.trainer_barrier();
+  if (comm.am_trainer_master()) {
+    comm.all_gather<EvalType>(score, score_list, comm.get_intertrainer_comm());
+  }
+  // Communicate trainer score list from trainer master to other procs in
+  // trainer
+  comm.trainer_broadcast(comm.get_trainer_master(),
+                         score_list.data(),
+                         num_trainers);
+
+  // Find winning trainer (in sample)
+  El::Int winner_id =
+    std::distance(score_list.begin(),
+                  std::max_element(score_list.begin(), score_list.end()));
+
+  // Find oldest trainer - cycle through trainer ids
+  El::Int oldest_id = step % num_trainers;
+
+  // DEBUG
+  std::cout << "Winner - " << winner_id << ", Oldest - " << oldest_id
+            << std::endl;
+
+  if (trainer_id == winner_id) {
+
+    if (winner_id != oldest_id) {
+      auto model_string = pack(m);
+      if (comm.am_trainer_master()) {
+        send_string(comm, model_string, oldest_id);
+        std::cout << "In Reg Evo step " << step << ", trainer " << trainer_id
+                  << " with score " << score_list[trainer_id];
+        std::cout << " sends model to trainer " << oldest_id << std::endl;
+      }
+    }
+  }
+
+  if (trainer_id == oldest_id) {
+
+    auto partner_model_ptr = m.copy_model();
+    auto& partner_model = *partner_model_ptr;
+
+    if (winner_id != oldest_id) {
+      std::string rcv_str;
+      if (comm.am_trainer_master()) {
+        rcv_str = recv_string(comm, winner_id);
+        std::cout << "In Reg Evo step " << step << ", trainer " << trainer_id;
+        std::cout << " receives model from trainer " << winner_id << std::endl;
+      }
+
+      unpack(partner_model, rcv_str);
+    }
+
+    // Mutating oldest model
+    m_mutate_algo->mutate(partner_model, step);
```

Comment:
I don't think the oldest model is being replaced by the winning model:
```suggestion

    if (winner_id != oldest_id) {
      std::string rcv_str;
      if (comm.am_trainer_master()) {
        rcv_str = recv_string(comm, winner_id);
        std::cout << "In Reg Evo step " << step << ", trainer " << trainer_id;
        std::cout << " receives model from trainer " << winner_id << std::endl;
      }

      unpack(m, rcv_str);
    }

    // Mutating oldest model
    m_mutate_algo->mutate(m, step);
```

Reply:
Solved